components/servo/message_port.rs

```diff
@@ -0,0 +1,163 @@
+    fn post_message_received(&self, _webview: WebView, _message_port: MessagePort, _data: JSValue) {
+    }
+}
+
+enum MessagePortState {
+    Ready {
+        id: MessagePortId,
+        webview_id: WebViewId,
+        entangled_id: MessagePortId,
+        constellation_proxy: ConstellationProxy,
+        delegate: Option<Rc<dyn MessagePortDelegate>>,
+    },
+    Detached,
+}
+
+#[derive(Clone)]
+pub struct MessagePort(Rc<RefCell<MessagePortState>>);
+
+#[derive(Debug, PartialEq)]
+pub struct MessagePortDetached;
+
+impl MessagePort {
+    fn new(
+        constellation_proxy: ConstellationProxy,
+        webview_id: WebViewId,
+        port_id: MessagePortId,
+        entangled_id: MessagePortId,
+    ) -> MessagePort {
+        MessagePort(Rc::new(RefCell::new(MessagePortState::Ready {
+            constellation_proxy,
+            webview_id,
+            id: port_id,
+            entangled_id,
+            delegate: None,
+        })))
+    }
+
+    pub(crate) fn new_entangled(
+        constellation_proxy: ConstellationProxy,
+        embedder_proxy: EmbedderProxy,
+        webview_id: WebViewId,
+    ) -> (MessagePort, MessagePort) {
+        let (port_control_sender, port_control_receiver) =
+            ipc::channel().expect("ipc channel failure");
+        ROUTER.add_typed_route(
+            port_control_receiver,
+            Box::new(move |message| match message.unwrap() {
+                MessagePortMsg::CompleteTransfer(..) => unreachable!(),
+                MessagePortMsg::CompletePendingTransfer(..) => unreachable!(),
+                MessagePortMsg::CompleteDisentanglement(..) => unreachable!(),
+                MessagePortMsg::NewTask(message_port_id, task) => {
+                    let data = match task.data {
+                        PostMessageData::StructuredClone(..) => unreachable!(),
+                        PostMessageData::Serialized(data) => data,
+                    };
+                    let _ = embedder_proxy.send(EmbedderMsg::PostMessage(
+                        webview_id,
+                        message_port_id,
+                        data,
+                    ));
+                },
+            }),
+        );
+        let first_id = MessagePortId::new();
+        let second_id = MessagePortId::new();
+        let router_id = MessagePortRouterId::new();
+        constellation_proxy.send(EmbedderToConstellationMessage::CreateEntangledMessagePorts(
+            router_id,
+            port_control_sender,
+            first_id,
+            second_id,
+        ));
+
+        let first = MessagePort::new(constellation_proxy.clone(), webview_id, first_id, second_id);
+        let second = MessagePort::new(constellation_proxy, webview_id, second_id, first_id);
+
+        (first, second)
+    }
+
+    /// Post the provided [JSValue] to the MessagePort that is entangled with this one.
+    pub fn post_message(&self, js_value: JSValue) -> Result<(), MessagePortDetached> {
```

Comment:
Here also, would be good to match https://html.spec.whatwg.org/multipage/web-messaging.html#message-port-post-message-steps

in particular, cannot transfer source or target port, things like that.